src/component_selector.jl

```diff
@@ -0,0 +1,295 @@
+"""`ComponentSelector`s that may refer to multiple components."""
+abstract type PluralComponentSelector <: ComponentSelector end
+
+#=
+`DynamicallyGroupedComponentSelector` extension notes:
+One MAY subtype this and have a `groupby::Union{Symbol, Function}` field to get an automatic
+implementation of `get_groups`.
+=#
+"`PluralComponentSelector`s whose grouping is determined by a `groupby` field."
+abstract type DynamicallyGroupedComponentSelector <: PluralComponentSelector end
+
+# COMMON COMPONENTSELECTOR INFRASTRUCTURE
+"Canonical way to turn an `InfrastructureSystemsComponent` subtype into a unique string."
+subtype_to_string(subtype::Type{<:InfrastructureSystemsComponent}) =
+    strip_module_name(subtype)
+
+"""
+Canonical way to turn an `InfrastructureSystemsComponent` specification/instance into a
+unique-per-system string.
+"""
+component_to_qualified_string(
+    component_subtype::Type{<:InfrastructureSystemsComponent},
+    component_name::AbstractString,
+) = subtype_to_string(component_subtype) * COMPONENT_NAME_DELIMETER * component_name
+component_to_qualified_string(component::InfrastructureSystemsComponent) =
+    component_to_qualified_string(typeof(component), get_name(component))
+
+const VALID_GROUPBY_KEYWORDS = [:all, :each]
+
+# Ideally we could leave system-like arguments untyped for maximum extensibility, but
+# because only PSY.get_components, not IS.get_components, is defined for PSY.System, we need
+# to be able to easily override these methods. See
+# https://github.com/NREL-Sienna/InfrastructureSystems.jl/issues/388
+const SYSTEM_LIKE = Union{Components, SystemData}
+
+"""
+Helper function to check that the `groupby` argument is valid. Passes it through if so,
+errors if not.
+"""
+validate_groupby(groupby::Symbol) =
+    if (groupby in VALID_GROUPBY_KEYWORDS)
+        groupby
+    else
+        throw(ArgumentError("groupby must be one of $VALID_GROUPBY_KEYWORDS or a function"))
+    end
+validate_groupby(groupby::Function) = groupby  # Don't try to validate functions for now
+
+# Generic implementations/generic docstrings for simple functions with many methods
+"""
+Factory function to create the appropriate subtype of `ComponentSelector` given the
+arguments. Users should call this rather than manually constructing `ComponentSelector`
+subtypes.
+"""
+function make_selector end
+
+"""
+Get the default name for the `ComponentSelector`, constructed automatically from what the
+`ComponentSelector` contains. Particularly for complex `ComponentSelector`s, this may not
+always be very concise or informative, so in these cases constructing the
+`ComponentSelector` with a custom name is recommended.
+"""
+function default_name end
+
+# Override this if you define a ComponentSelector subtype with no name field
+"""
+Get the name of the `ComponentSelector`. This is either the default name or a custom name
+passed in at creation time.
+"""
+get_name(e::ComponentSelector) = (e.name !== nothing) ? e.name : default_name(e)
```

Comment:
It would slightly simplify the code (though would require manual constructors…). I remember thinking about this a long time ago, I'll try to reconstruct why I did it this way…

Reply:
Okay, I bit the bullet and did manual constructors. `default_name` is eliminated now